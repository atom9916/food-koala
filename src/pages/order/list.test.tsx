import Orderlist from './list';
import { render, screen } from '@testing-library/react';
import useSWR, { Middleware, SWRConfig, SWRResponse } from 'swr';
import renderer from 'react-test-renderer';
import { useRouter } from 'next/router';

const fetchMockError = () => {
  const error = new Error('Failed to fetch data');
  return new Promise((rejects) => {
    rejects({
      ok: false,
      status: 400,
      json: async () => {
        throw error;
      },
    });
  });
};

describe('OrderList', () => {
  beforeAll(() => {
    jest.mock('next/router', () => ({
      useRouter: jest.fn().mockReturnValue({
        replace: jest.fn(),
      }),
    }));
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('renders loding when data is not fetched', () => {
    render(<Orderlist />);
    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('renders Error when fetching data is failed', async () => {
    global.fetch = jest.fn().mockImplementation(fetchMockError);

    const { asFragment } = render(<Orderlist />);

    expect(await screen.findByText('Error...')).toBeInTheDocument();
    expect(fetch).toHaveBeenCalledWith('/api/get_cart_items');
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders OrderList correctly when the cart is empty', async () => {
    const testMiddleware: Middleware = () => {
      return (): SWRResponse<any, any> => {
        const mockData: any = [];
        return {
          data: mockData,
          error: undefined,
          mutate: (_) => Promise.resolve(),
          isValidating: false,
          isLoading: false,
        };
      };
    };

    // Orderlistをレンダリング
    const tree = renderer
      .create(
        <SWRConfig value={{ use: [testMiddleware] }}>
          <Orderlist />
        </SWRConfig>
      )
      .toJSON();

    //エラーメッセージが画面に描写されているか確認
    const errorMessage = await screen.findByText(
      '※カートに商品がないため、購入できません'
    );
    expect(errorMessage).toBeInTheDocument();

    //リンクが画面に描写されているか確認
    const link = screen.getByRole('link', { name: 'メニュー一覧へ' });
    expect(link).toBeInTheDocument();

    //fetchの引数が正しいか確認
    expect(fetch).toHaveBeenCalledWith('/api/get_cart_items');

    //スナップショットテスト
    expect(tree).toMatchSnapshot();
  });
});
