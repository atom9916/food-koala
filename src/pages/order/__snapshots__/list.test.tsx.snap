// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderList renders Error when fetching data is failed 1`] = `
<DocumentFragment>
  <div>
    Error...
  </div>
</DocumentFragment>
`;

exports[`OrderList renders OrderList correctly when the cart is empty 1`] = `
<DocumentFragment>
  <div>
    Error...
  </div>
</DocumentFragment>
`;
